src/shm/shm.c

```diff
@@ -0,0 +1,256 @@
+#include "shm.h"
+#include <sys/mman.h>
+#include <sys/file.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+
+// Tries to access the coordination shared memory area or creates it if
+// it doesn't yet exist
+mmap_creat_ret_t access_or_create_coord(char* passwd){
+    char* path = malloc(strlen("/ocp_sync")+strlen(passwd)+1);
+    strcpy(path, "/ocp_sync");
+    strcat(path, passwd);
+    int fdm = shm_open(path, O_CREAT | O_EXCL | O_RDWR | O_TRUNC, 0660);
+    printf("opn_sync %d %d\n", fdm, errno);
+    if (fdm == -1){
+        if (errno == EEXIST){
+            fdm = shm_open(path, O_RDWR, 0);
+            coord_struct* str = (coord_struct*) mmap(NULL, sizeof(coord_struct), PROT_READ | PROT_WRITE, MAP_SHARED, fdm, 0);
+            printf("mmp sync %p %d\n", str, errno);
+            return (mmap_creat_ret_t){str, fdm, path};
+        } else {
+            err_and_leave("Error when accessing shared memory", 5);
+        }
+    }
+    ftruncate(fdm, sizeof(coord_struct));
+    coord_struct* str = /*malloc(sizeof(coord_struct));/*/ (coord_struct*) mmap(NULL, sizeof(coord_struct), PROT_READ | PROT_WRITE, MAP_SHARED, fdm, 0);
+    // TODO: init all locks and the like here, returning will relinquish the lock and others might access
+    init_mutex(&(str->lock));
+    init_cond_pack(&(str->reader_ready));
+    init_cond_pack(&(str->writer_ready));
+    str->abort = 0;
+
+    return (mmap_creat_ret_t){str, fdm, path};
+}
+
+char* path_copy(char* passwd){
+    char* path = malloc(2+strlen(passwd)+strlen("_copy"));
+    strcpy(path, "/");
+    strcat(path, passwd);
+    strcat(path,"_copy");
+    return path;
+}
+
+mmap_creat_ret_t init_copy_area(char* passwd, size_t width, size_t* mmap_size){
+    int fdm;
+    char *path = path_copy(passwd);
+    { // Create file for shared_memory
+        fdm = shm_open(path, O_CREAT | O_EXCL | O_RDWR | O_TRUNC, 0660);
+        printf("opn_cp_shm %d %d\n", fdm, errno);
+        if (fdm == -1){
+            err_and_leave("Error when creating file-specific sharedmemory", 5);
+        }
+    }
+    size_t m_sz = sizeof(copy_struct) + (2*width);
+    *mmap_size = m_sz;
```

Comment:
The size allocated to the copy struct is not known at compile time since it depends on width (in this case width is a constant but I plan to take width as a runtime parameter with a further CLI argument)

The receiver process needs to know how much space to allocate for the memory map later, I could either have the receiver perform the same computation or (as I have opted) I save into the coordination structure variable the value used in this function

Reply:
Still, `m_sz` is lost at the end of the function, no ?

Reply:
Yes, but the value has been copied into mmap_size (which is just a pointer to a field of the coordination structure) so the value survives

Reply:
Right, it is not mmap_size = &m_size
The line is gone anyway.